server/src/main/kotlin/io/provenance/objectstore/gateway/service/StreamEventHandlerService.kt

```diff
@@ -15,29 +23,127 @@ class StreamEventHandlerService(
     private val accountAddresses: Set<String>,
     private val scopePermissionsRepository: ScopePermissionsRepository,
     private val pbClient: PbClient,
+    private val provenanceProperties: ProvenanceProperties,
 ) {
     private companion object : KLogging()
 
+    /**
+     * The core entrypoint for this service.  Consumes an event and attempts to parse its values as gateway-mandated
+     * values for taking specified actions.
+     *
+     * @param event Any event emitted by a wasm smart contract that may or may not conform to the specified event structure
+     * of object store gateway.
+     */
     fun handleEvent(event: TxEvent) {
-        handleEvent(AssetClassificationEvent(event))
+        // Try first to intercept incoming events as Gateway events, but fallback to asset classification if the event
+        // does not fit the expected gateway event structure
+        GatewayEvent.fromEventOrNull(event)?.also { gatewayEvent ->
+            when (gatewayEvent.eventType) {
+                GatewayExpectedEventType.ACCESS_GRANT -> handleGatewayGrant(gatewayEvent)
+                GatewayExpectedEventType.ACCESS_REVOKE -> handleGatewayRevoke(gatewayEvent)
+            }
+        } ?: handleAssetClassificationEvent(AssetClassificationEvent(event))
+    }
+
+    /**
+     * Uses a successfully-parsed gateway grant event to grant access to a scope's underlying records if all validation
+     * for the event's formation passes.
+     *
+     * @param gatewayEvent A wasm-emitted event that uses the gateway access grant event type.
+     */
+    private fun handleGatewayGrant(gatewayEvent: GatewayEvent) {
+        val scopeResponse = lookupScope(gatewayEvent.scopeAddress)
+        // Ensure that the referenced scope is owned in some way by one of the gateway's registered keys.
+        // This ensures that properly-stored scopes will be able to be successfully decrypted in future requests by
+        // the granted address.
+        val granterAddress = findRegisteredScopeOwnerAddress(scopeResponse) ?: run {
+            logger.info("Skipping $gatewayEvent not owned by any registered granters")
+            return
+        }
+        // Ensure that the scope's value owner signed for the transaction that emitted this wasm event.  This ensures
+        // that this access grant has been safely requested and should be made
+        getSignerAddressesForTx(gatewayEvent.txHash).also { signerAddresses ->
+            if (scopeResponse.scope.scope.valueOwnerAddress !in signerAddresses) {
```

Comment:
We should probably have a TODO in here about adding a fallback check for an authz contract execute grant from the valueOwnerAddress to any of the signerAddresses... I think we will need that, right? In the case where someone delegates the classification of their scopes for them, for instance?

Reply:
I totally forgot about that and you're absolutely right.  I'll pop that TODO on there

Reply:
Done did the thing 👍